src/ui/pages/SwapForm/SwapForm.tsx

```diff
@@ -287,6 +306,75 @@ export function SwapFormComponent() {
   const sendTxBtnRef = useRef<SendTxBtnHandle | null>(null);
   const approveTxBtnRef = useRef<SendTxBtnHandle | null>(null);
 
+  const txToCheck =
+    quotesData.done && !enough_allowance
+      ? approvalTransaction
+      : swapTransaction;
+
+  const { data: networkNonce } = useQuery({
+    suspense: false,
+    queryKey: ['uiGetBestKnownTransactionCount', address, chain, networks],
+    queryFn: async () => {
+      if (!chain || !networks) {
+        return null;
+      }
+      const result = await uiGetBestKnownTransactionCount({
+        address,
+        chain,
+        networks,
+        defaultBlock: 'pending',
+      });
+      return result.value;
+    },
+  });
+
+  const txNonce = txToCheck && 'nonce' in txToCheck ? txToCheck.nonce : null;
+
+  const isDeviceWallet = wallet && isDeviceAccount(wallet);
+  const USE_PAYMASTER_FEATURE = !isDeviceWallet;
+
+  const configuration = useStore(swapView.store.configuration);
+  const userNonce = configuration.nonce;
+  const nonce = userNonce ?? txNonce ?? networkNonce ?? undefined;
+  const gas = txToCheck ? getGas(txToCheck) : null;
+  const eligibilityParams:
+    | null
+    | Parameters<ZerionApiClient['paymasterCheckEligibility']>[0] =
+    USE_PAYMASTER_FEATURE && txToCheck && txToCheck.to && nonce != null && gas
+      ? {
+          chainId: txToCheck.chainId,
+          from: address,
+          to: txToCheck.to,
+          value: txToCheck.value,
+          data: txToCheck.data,
+          gas,
+          nonce,
+        }
+      : null;
+  const eligibiliteQuery = useTxEligibility(eligibilityParams, {
+    enabled:
+      USE_PAYMASTER_FEATURE &&
+      Boolean(network?.supports_sponsored_transactions),
+  });
+
+  const paymasterEligible = Boolean(eligibiliteQuery.data?.data.eligible);
+
+  const configureTransactionToBeSigned = useEvent(
+    (tx: IncomingTransactionWithChainId) => {
+      invariant(chain && networks, 'Not ready to prepare the transaction');
+      const configuration = swapView.store.configuration.getState();
+      let txToSign = applyConfiguration(tx, configuration, gasPrices);
+      if (paymasterEligible && txToSign.nonce == null) {
+        txToSign = { ...txToSign, nonce };
+        invariant(txToSign.nonce, 'Nonce is required for a paymaster tx');
+      }
+      return { ...txToSign, from: address };
```

Comment:
why don't we add chain id here anymore?

Reply:
Because thankfully it's guaranteed at this point